src/context/Iterator.cpp

```diff
@@ -388,6 +388,208 @@ size_t JoinIter::buildIndexFromJoinIter(const JoinIter* iter, size_t segIdx) {
+        LOG(ERROR) << status;
+        clear();
+        return;
+    }
+    for (auto& row : ds.rows) {
+        rows_.emplace_back(&row);
+    }
+    iter_ = rows_.begin();
+}
+
+Status PropIter::makeDataSetIndex(const DataSet& ds) {
+    dsIndex_.ds = &ds;
+    auto& colNames = ds.colNames;
+    for (size_t i = 0; i < colNames.size(); ++i) {
+        dsIndex_.colIndices.emplace(colNames[i], i);
+        auto& colName = colNames[i];
+        if (colName.find(".") != std::string::npos) {
+            NG_RETURN_IF_ERROR(buildPropIndex(colName, i));
+        }
+    }
+    return Status::OK();
+}
+
+Status PropIter::buildPropIndex(const std::string& props, size_t columnId) {
+    std::vector<std::string> pieces;
+    folly::split(".", props, pieces);
+    if (UNLIKELY(pieces.size() != 2)) {
+        return Status::Error("Bad column name format: %s", props.c_str());
+    }
+    std::string name = pieces[0];
+    auto& propsMap = dsIndex_.propsMap;
+    if (propsMap.find(name) != propsMap.end()) {
+        auto& propIndex = propsMap[name];
+        propIndex.colIndexs.emplace_back(columnId);
+        propIndex.propIndices.emplace(pieces[1], columnId);
+    } else {
+        PropIndex propIndex;
+        propIndex.colIndexs.emplace_back(columnId);
+        propIndex.propIndices.emplace(pieces[1], columnId);
+        propsMap.emplace(name, std::move(propIndex));
+    }
+    return Status::OK();
+}
+
+const Value& PropIter::getColumn(const std::string& col) const {
+    if (!valid()) {
+        return Value::kNullValue;
+    }
+
+    auto& logicalRow = *iter_;
+    auto index = dsIndex_.colIndices.find(col);
+    if (index == dsIndex_.colIndices.end()) {
+        return Value::kNullValue;
+    } else {
+        DCHECK_LT(index->second, logicalRow.row_->values.size());
+        return logicalRow.row_->values[index->second];
+    }
+}
+
+const Value& PropIter::getProp(const std::string& name, const std::string& prop) const {
+    if (!valid()) {
+        return Value::kNullValue;
+    }
+    auto& row = *(iter_->row_);
+    auto& propsMap = dsIndex_.propsMap;
+    auto index = propsMap.find(name);
+    if (index == propsMap.end()) {
+        return Value::kEmpty;
+    }
+
+    auto propIndex = index->second.propIndices.find(prop);
+    if (propIndex == index->second.propIndices.end()) {
+        VLOG(1) << "No prop found : " << prop;
+        return Value::kNullValue;
+    }
+    auto colId = propIndex->second;
+    DCHECK_GT(row.size(), colId);
+    return row[colId];
+}
+
+Value PropIter::getVertex() const {
+    if (!valid()) {
+        return Value::kNullValue;
+    }
+
+    auto vidVal = getColumn(nebula::kVid);
+    if (!vidVal.isStr()) {
+        return Value::kNullValue;
+    }
+    Vertex vertex;
+    vertex.vid = vidVal.getStr();
+    auto& tagPropsMap = dsIndex_.propsMap;
+    bool isVertexProps = true;
+    for (auto& tagProp : tagPropsMap) {
+        auto& row = *(iter_->row_);
+        auto& colIndexs = tagProp.second.colIndexs;
```

Comment:
I think the `colIndexs` is useless. propIndices has included it.

Reply:
done